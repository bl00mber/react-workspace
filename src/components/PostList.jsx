import React from 'react';
import PropTypes from 'prop-types';

const PostList = props => (
  <div>
    {props.isFetching && <p>Fetching...</p>}
    {!props.isFetching && props.error && <p>Error: {props.error.description}</p>}
    <ul>
      {
        props.posts.map(post =>
          <li key={post.id}>{post.description}</li>)
      }
    </ul>
    <button onClick={() => props.onFetchPosts(true)}>Fetch Posts OK</button>
    <button onClick={() => props.onFetchPosts(false)}>Fetch Posts Errors</button>
  </div>
);

PostList.defaultProps = {
  error: null,
};

PostList.propTypes = {
  isFetching: PropTypes.bool.isRequired,
  error: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  posts: PropTypes.arrayOf(PropTypes.object).isRequired,
  onFetchPosts: PropTypes.func.isRequired,
};

export default PostList;
