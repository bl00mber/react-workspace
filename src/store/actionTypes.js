/*
 * action types
 */

export const ADD_TODO = 'ADD_TODO';
export const TOGGLE_TODO = 'TOGGLE_TODO';
export const SET_VISIBILITY_FILTER = 'SET_VISIBILITY_FILTER';
export const INCREMENT_COUNTER = 'INCREMENT_COUNTER';
export const DECREMENT_COUNTER = 'DECREMENT_COUNTER';
export const TOGGLE_FLAG = 'TOGGLE_FLAG';
export const SET_CELSIUS = 'SET_CELSIUS';
export const SET_FAHRENHEIT = 'SET_FAHRENHEIT';
export const FETCH_POSTS_REQUEST = 'FETCH_POSTS_REQUEST';
export const FETCH_POSTS_FAILURE = 'FETCH_POSTS_FAILURE';
export const FETCH_POSTS_SUCCESS = 'FETCH_POSTS_SUCCESS';
