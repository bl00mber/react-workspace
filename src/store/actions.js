import {
  ADD_TODO,
  TOGGLE_TODO,
  SET_VISIBILITY_FILTER,
  TOGGLE_FLAG,
  SET_CELSIUS,
  SET_FAHRENHEIT,
} from './actionTypes';

/*
 * other constants
 */

export const VisibilityFilters = {
  SHOW_ALL: 'SHOW_ALL',
  SHOW_COMPLETED: 'SHOW_COMPLETED',
  SHOW_ACTIVE: 'SHOW_ACTIVE',
};

/* --------------- */


export const addTodo = (id, text) => ({
  type: ADD_TODO,
  id,
  text,
});


export const toggleTodo = id => ({
  type: TOGGLE_TODO,
  id,
});


export const setVisibilityFilter = filter => ({
  type: SET_VISIBILITY_FILTER,
  filter,
});

export const toggleFlag = flag => ({
  type: TOGGLE_FLAG,
  flag,
});

export const setCelsius = temperature => ({
  type: SET_CELSIUS,
  temperature,
});

export const setFahrenheit = temperature => ({
  type: SET_FAHRENHEIT,
  temperature,
});
