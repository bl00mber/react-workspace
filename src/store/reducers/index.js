import { combineReducers } from 'redux';

import counter from './counter';
import todos from './todos';
import visibilityFilter from './visibilityFilter';
import toggle from './toggle';

export default combineReducers({
  counter,
  todos,
  visibilityFilter,
  toggle,
});
