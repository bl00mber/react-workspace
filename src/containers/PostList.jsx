import { connect } from 'react-redux';

import PostListComponent from '@/components/PostList';
import { fetchPostsAsync } from '@/store/actions';

const mapStateToProps = state => ({
  isFetching: state.posts.isFetching,
  error: state.posts.error,
  posts: state.posts.posts,
});

const mapDispatchToProps = dispatch => ({
  onFetchPosts: params => fetchPostsAsync(params)(dispatch),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(PostListComponent);
