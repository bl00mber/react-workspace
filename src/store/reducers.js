import { combineReducers } from 'redux';

import { VisibilityFilters } from './actions';
import {
  ADD_TODO,
  TOGGLE_TODO,
  SET_VISIBILITY_FILTER,
  INCREMENT_COUNTER,
  DECREMENT_COUNTER,
} from './actionTypes';

const { SHOW_ALL } = VisibilityFilters;

const counter = (state = 0, action) => {
  switch (action.type) {
    case INCREMENT_COUNTER:
      return state + 1;
    case DECREMENT_COUNTER:
      return state - 1;
    default:
      return state;
  }
};

const todos = (state = [], action) => {
  switch (action.type) {
    case ADD_TODO:
      return [
        ...state,
        {
          text: action.text,
          completed: false,
        },
      ];
    case TOGGLE_TODO:
      return state.map((todo, index) => {
        if (index === action.index) {
          return Object.assign({}, todo, {
            completed: !todo.completed,
          });
        }
        return todo;
      });
    default:
      return state;
  }
};

const visibilityFilter = (state = SHOW_ALL, action) => {
  switch (action.type) {
    case SET_VISIBILITY_FILTER:
      return action.filter;
    default:
      return state;
  }
};

export default combineReducers({
  counter,
  todos,
  visibilityFilter,
});
