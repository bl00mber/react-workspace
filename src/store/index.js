import { createStore } from 'redux';

import actions from './actions';
import actionTypes from './actionTypes';
import reducers from './reducers';

/**
 * This is a reducer, a pure function with (state, action) => state signature.
 * It describes how an action transforms the state into the next state.
 *
 * The shape of the state is up to you: it can be a primitive, an array, an object,
 * or even an Immutable.js data structure. The only important part is that you should
 * not mutate the state object, but return a new object if the state changes.
 *
 * In this example, we use a `switch` statement and strings, but you can use a helper that
 * follows a different convention (such as function maps) if it makes sense for your
 * project.
 */
const counter = (state = 0, action) => {
  switch (action.type) {
    case 'INCREMENT':
      return state + 1;
    case 'DECREMENT':
      return state - 1;
    default:
      return state;
  }
};

// Create a Redux store holding the state of your app.
// Its API is { subscribe, dispatch, getState }.
const isDev = process.env.NODE_ENV === 'development';
const hasReduxDevTools = window.__REDUX_DEVTOOLS_EXTENSION__; // eslint-disable-line
const reduxDevTools = hasReduxDevTools && window.__REDUX_DEVTOOLS_EXTENSION__()  // eslint-disable-line
const store = createStore(
  counter,
  hasReduxDevTools && isDev && reduxDevTools,
);

// You can use subscribe() to update the UI in response to state changes.
// Normally you'd use a view binding library (e.g. React Redux) rather than subscribe() directly.
// However it can also be handy to persist the current state in the localStorage.
// store.subscribe(() => console.log(store.getState()));

export {
  actions,
  actionTypes,
  reducers,
  store,
};

export default store;
