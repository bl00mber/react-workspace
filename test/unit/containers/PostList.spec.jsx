import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import PostListComponent from '@/components/PostList';
import PostList from '@/containers/PostList';
import { fetchPostsAsync } from '@/store/actions';

let returnFetchPostAsync;
jest.mock('@/store/actions', () => ({
  fetchPostsAsync: jest.fn(),
}));

const state = {
  posts: {
    isFetching: false,
    error: null,
    posts: [],
  },
};

let store;


beforeEach(() => {
  store = {
    getState: jest.fn().mockReturnValue(state),
    subscribe: jest.fn(),
    dispatch: jest.fn(),
  };

  fetchPostsAsync.mockReset();
  returnFetchPostAsync = jest.fn();
  fetchPostsAsync.mockReturnValue(returnFetchPostAsync);
});

describe('PostList', () => {
  test('should match enzyme shallow snapshot', () => {
    const props = {
      store,
    };
    const wrapper = shallow(<PostList {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should have correct props mapped', () => {
    const props = {
      store,
    };
    const wrapper = shallow(<PostList {...props} />);
    const {
      isFetching, error, posts, onFetchPosts,
    } = wrapper.find(PostListComponent).props();
    expect(isFetching).toBe(state.posts.isFetching);
    expect(error).toBe(state.posts.error);
    expect(posts).toBe(state.posts.posts);
    expect(onFetchPosts).toBeDefined();
    expect(typeof onFetchPosts).toBe('function');
  });

  test('should dispatch in the right way', () => {
    const props = {
      store,
    };
    const wrapper = shallow(<PostList {...props} />);
    const {
      onFetchPosts,
    } = wrapper.find(PostListComponent).props();

    onFetchPosts('params');
    expect(fetchPostsAsync).toHaveBeenCalledWith('params');
    expect(store.dispatch).not.toHaveBeenCalled();
    expect(returnFetchPostAsync).toHaveBeenCalledWith(store.dispatch);
  });
});
