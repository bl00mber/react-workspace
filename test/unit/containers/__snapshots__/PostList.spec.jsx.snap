// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostList should match enzyme shallow snapshot 1`] = `
<PostList
  error={null}
  isFetching={false}
  onFetchPosts={[Function]}
  posts={Array []}
  store={
    Object {
      "dispatch": [MockFunction],
      "getState": [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
      },
      "subscribe": [MockFunction] {
        "calls": Array [
          Array [
            [Function],
          ],
        ],
      },
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [MockFunction],
        "getState": [MockFunction] {
          "calls": Array [
            Array [],
            Array [],
          ],
        },
        "subscribe": [MockFunction] {
          "calls": Array [
            Array [
              [Function],
            ],
          ],
        },
      },
      "unsubscribe": undefined,
    }
  }
/>
`;
